proxy/src/auth/credentials.rs

```diff
@@ -153,10 +149,46 @@ impl<'a> ClientCredentials<'a> {
             user,
             project,
             cache_key,
+            peer_addr,
         })
     }
 }
 
+pub fn check_peer_addr_is_in_list(peer_addr: &IpAddr, ip_list: &Vec<String>) -> bool {
+    if ip_list.is_empty() {
+        return true;
+    }
+    for ip in ip_list {
+        // We expect that all ip addresses from control plane are correct.
+        // However, if some of them are broken, we still can check the others.
+        match check_ip(peer_addr, ip) {
+            Ok(result) => {
+                if result {
+                    return true;
+                }
+            }
+            Err(err) => warn!("Cannot parse ip: {}; err: {}", ip, err),
+        }
+    }
+    false
+}
+
+fn check_ip(ip: &IpAddr, pattern: &str) -> anyhow::Result<bool> {
+    if pattern.contains('/') {
+        let subnet: ipnet::IpNet = pattern.parse()?;
+        return Ok(subnet.contains(ip));
+    }
+    if let Some((start, end)) = pattern.split_once('-') {
+        // let range: ipnet::IpAddrRange = range.parse()?;
+        let start: IpAddr = start.parse()?;
+        let end: IpAddr = end.parse()?;
+        Ok(start <= *ip && *ip <= end)
+    } else {
+        let addr: IpAddr = pattern.parse()?;
+        Ok(addr.eq(ip))
+    }
+}
```

Comment:
I would quite like to split this in two.

1: `parse_ip_pattern(&str) -> Result<IpPattern>;` probably as a `FromStr` impl and preferably as a `Deserialize` impl too.
2: `IpPattern::check(IpAddr) -> bool;`